# Data Science Project : Apple iPhone Sales Analysis


In [25]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

In [26]:
data = pd.read_csv("apple_products.csv")

In [27]:
print(data.head())
print(data.columns)

                               Product Name  \
0         APPLE iPhone 8 Plus (Gold, 64 GB)   
1  APPLE iPhone 8 Plus (Space Grey, 256 GB)   
2      APPLE iPhone 8 Plus (Silver, 256 GB)   
3           APPLE iPhone 8 (Silver, 256 GB)   
4             APPLE iPhone 8 (Gold, 256 GB)   

                                         Product URL  Brand  Sale Price  \
0  https://www.flipkart.com/apple-iphone-8-plus-g...  Apple       49900   
1  https://www.flipkart.com/apple-iphone-8-plus-s...  Apple       84900   
2  https://www.flipkart.com/apple-iphone-8-plus-s...  Apple       84900   
3  https://www.flipkart.com/apple-iphone-8-silver...  Apple       77000   
4  https://www.flipkart.com/apple-iphone-8-gold-2...  Apple       77000   

     Mrp  Discount Percentage  Number Of Ratings  Number Of Reviews  \
0  49900                    0               3431                356   
1  84900                    0               3431                356   
2  84900                    0               3431     

In [28]:
print(data.isnull().sum())

Product Name           0
Product URL            0
Brand                  0
Sale Price             0
Mrp                    0
Discount Percentage    0
Number Of Ratings      0
Number Of Reviews      0
Upc                    0
Star Rating            0
Ram                    0
dtype: int64


In [29]:
data.describe()

Sale Price            Mrp  Discount Percentage  Number Of Ratings  \
count      62.000000      62.000000            62.000000          62.000000   
mean    80073.887097   88058.064516             9.951613       22420.403226   
std     34310.446132   34728.825597             7.608079       33768.589550   
min     29999.000000   39900.000000             0.000000         542.000000   
25%     49900.000000   54900.000000             6.000000         740.000000   
50%     75900.000000   79900.000000            10.000000        2101.000000   
75%    117100.000000  120950.000000            14.000000       43470.000000   
max    140900.000000  149900.000000            29.000000       95909.000000   

       Number Of Reviews  Star Rating  
count          62.000000    62.000000  
mean         1861.677419     4.575806  
std          2855.883830     0.059190  
min            42.000000     4.500000  
25%            64.000000     4.500000  
50%           180.000000     4.600000  
75%          3331.000000     4.600000  
max          8161.000000     4.700000

In [30]:
highest_rated = data.sort_values(by=['Star Rating'], ascending=False)
highest_rated

Product Name  \
15             APPLE iPhone 11 Pro Max (Gold, 64 GB)   
20   APPLE iPhone 11 Pro Max (Midnight Green, 64 GB)   
17       APPLE iPhone 11 Pro Max (Space Grey, 64 GB)   
16  APPLE iPhone 11 Pro Max (Midnight Green, 256 GB)   
14            APPLE iPhone 11 Pro Max (Gold, 256 GB)   
..                                               ...   
52                    APPLE iPhone SE (White, 64 GB)   
53                   APPLE iPhone SE (Black, 128 GB)   
57                    APPLE iPhone SE (Black, 64 GB)   
54                   APPLE iPhone SE (White, 128 GB)   
55                     APPLE iPhone SE (Red, 128 GB)   

                                          Product URL  Brand  Sale Price  \
15  https://www.flipkart.com/apple-iphone-11-pro-m...  Apple      117100   
20  https://www.flipkart.com/apple-iphone-11-pro-m...  Apple      117100   
17  https://www.flipkart.com/apple-iphone-11-pro-m...  Apple      117100   
16  https://www.flipkart.com/apple-iphone-11-pro-m...  Apple      131900   
14  https://www.flipkart.com/apple-iphone-11-pro-m...  Apple      131900   
..                                                ...    ...         ...   
52  https://www.flipkart.com/apple-iphone-se-white...  Apple       29999   
53  https://www.flipkart.com/apple-iphone-se-black...  Apple       34999   
57  https://www.flipkart.com/apple-iphone-se-black...  Apple       29999   
54  https://www.flipkart.com/apple-iphone-se-white...  Apple       34999   
55  https://www.flipkart.com/apple-iphone-se-red-1...  Apple       34999   

       Mrp  Discount Percentage  Number Of Ratings  Number Of Reviews  \
15  117100                    0               1078                101   
20  117100                    0               1078                101   
17  117100                    0               1078                101   
16  131900                    0               1078                101   
14  131900                    0               1078                101   
..     ...                  ...                ...                ...   
52   39900                   24              95807               8154   
53   44900                   22              95909               8161   
57   39900                   24              95909               8161   
54   44900                   22              95807               8154   
55   44900                   22              95909               8161   

                 Upc  Star Rating   Ram  
15  MOBFKCTSAPAYNSGG          4.7  4 GB  
20  MOBFKCTSRYPAQNYT          4.7  4 GB  
17  MOBFKCTSKDMKCGQS          4.7  4 GB  
16  MOBFKCTSCAAKGQV7          4.7  4 GB  
14  MOBFKCTS7HCHSPFH          4.7  4 GB  
..               ...          ...   ...  
52  MOBFWQ6BGWDVGF3E          4.5  2 GB  
53  MOBFWQ6BHUEVZPXD          4.5  2 GB  
57  MOBFWQ6BR3MK7AUG          4.5  4 GB  
54  MOBFWQ6BJEHMUUZY          4.5  2 GB  
55  MOBFWQ6BJTVFKPEJ          4.5  2 GB  

[62 rows x 11 columns]

In [31]:
iphones = highest_rated["Product Name"].value_counts()
label = iphones.index
label


Index(['APPLE iPhone 11 Pro Max (Gold, 64 GB)',
       'APPLE iPhone 11 Pro Max (Midnight Green, 64 GB)',
       'APPLE iPhone 11 Pro Max (Space Grey, 64 GB)',
       'APPLE iPhone 11 Pro Max (Midnight Green, 256 GB)',
       'APPLE iPhone 11 Pro Max (Gold, 256 GB)',
       'APPLE iPhone 8 Plus (Gold, 64 GB)',
       'Apple iPhone XR (Black, 128 GB) (Includes EarPods, Power Adapter)',
       'Apple iPhone XR (Coral, 128 GB) (Includes EarPods, Power Adapter)',
       'Apple iPhone XR ((PRODUCT)RED, 128 GB) (Includes EarPods, Power Adapter)',
       'APPLE iPhone 8 Plus (Space Grey, 256 GB)',
       'APPLE iPhone 8 Plus (Silver, 256 GB)',
       'APPLE iPhone 8 Plus (Silver, 64 GB)',
       'APPLE iPhone XS Max (Silver, 64 GB)',
       'APPLE iPhone 8 Plus (Space Grey, 64 GB)',
       'APPLE iPhone 11 Pro (Space Grey, 256 GB)',
       'APPLE iPhone 12 Pro Max (Pacific Blue, 256 GB)',
       'APPLE iPhone 12 (White, 128 GB)',
       'Apple iPhone XR (White, 128 GB) (Includes EarPods, Powe

In [32]:
counts = highest_rated["Number Of Ratings"]
figure = px.bar(highest_rated, x=label, y=counts, title="Number of Ratings of Different Iphones")
figure.show()

In [33]:
iphones = highest_rated["Product Name"].value_counts()
label = iphones.index
counts = highest_rated["Number Of Reviews"]
figure = px.bar(highest_rated, x=label, y=counts, title="Number of Reviews of Different Iphones")
figure.show()

In [34]:
figure = px.scatter(data_frame = data, x='Number Of Ratings', y='Sale Price', size='Discount Percentage', title="Relationship between Sale Price, Number of Ratings and Discount Percentage")
figure.show()

In [35]:
figure = px.scatter(
    data_frame=data,
    x="Number Of Ratings",
    y="Discount Percentage",
    size="Sale Price",
    trendline="ols",
    title="Relationship between Discount Percentage and Number Of Ratings"
)

figure.show()

In [36]:
most_expensive = data.loc[data['Sale Price'].idxmax()]
least_expensive = data.loc[data['Sale Price'].idxmin()]

print("Most Expensive Product:")
print(most_expensive)

print("\nLeast Expensive Product:")
print(least_expensive)


Most Expensive Product:
Product Name                        APPLE iPhone 12 Pro (Silver, 512 GB)
Product URL            https://www.flipkart.com/apple-iphone-12-pro-s...
Brand                                                              Apple
Sale Price                                                        140900
Mrp                                                               149900
Discount Percentage                                                    6
Number Of Ratings                                                    542
Number Of Reviews                                                     42
Upc                                                     MOBFWBYZ5UY6ZBVA
Star Rating                                                          4.5
Ram                                                                 4 GB
Name: 24, dtype: object

Least Expensive Product:
Product Name                              APPLE iPhone SE (White, 64 GB)
Product URL            https://www.flipkart.com/ap

# Additional features


## Analyze Discount Percentage

### Subtask:
Calculate and display the average discount percentage across all products.


**Reasoning**:
To calculate the average discount percentage, I need to get the mean of the 'Discount Percentage' column from the 'data' DataFrame.



In [37]:
average_discount = data['Discount Percentage'].mean()
print(f"Average Discount Percentage: {average_discount:.2f}%")

Average Discount Percentage: 9.95%


## Identify Most Rated Products

### Subtask:
Identify and display the top 10 iPhones based on the 'Number Of Ratings'.


**Reasoning**:
To identify the top 10 iPhones by 'Number Of Ratings', I will first sort the 'data' DataFrame by this column in descending order, then select the top 10 products and display their names and number of ratings.



In [38]:
top_10_rated = data.sort_values(by='Number Of Ratings', ascending=False).head(10)
print("Top 10 iPhones by Number Of Ratings:")
print(top_10_rated[['Product Name', 'Number Of Ratings']])

Top 10 iPhones by Number Of Ratings:
                                         Product Name  Number Of Ratings
23  Apple iPhone SE (White, 256 GB) (Includes EarP...              95909
55                      APPLE iPhone SE (Red, 128 GB)              95909
57                     APPLE iPhone SE (Black, 64 GB)              95909
53                    APPLE iPhone SE (Black, 128 GB)              95909
52                     APPLE iPhone SE (White, 64 GB)              95807
54                    APPLE iPhone SE (White, 128 GB)              95807
11  Apple iPhone XR (Coral, 128 GB) (Includes EarP...              79582
12  Apple iPhone XR (Black, 128 GB) (Includes EarP...              79512
13  Apple iPhone XR (White, 128 GB) (Includes EarP...              79512
10  Apple iPhone XR (Black, 64 GB) (Includes EarPo...              79512


## Identify Most Reviewed Products

### Subtask:
Identify and display the top 10 iPhones based on the 'Number Of Reviews'.


**Reasoning**:
To identify the top 10 iPhones by 'Number Of Reviews', I will first sort the 'data' DataFrame by this column in descending order, then select the top 10 products and display their names and number of reviews.

**Reasoning**:
To identify the top 10 iPhones by 'Number Of Reviews', I will sort the DataFrame, select the top 10 rows, and display the product name and number of reviews.



In [39]:
top_10_reviewed = data.sort_values(by='Number Of Reviews', ascending=False).head(10)
print("Top 10 iPhones by Number Of Reviews:")
print(top_10_reviewed[['Product Name', 'Number Of Reviews']])

Top 10 iPhones by Number Of Reviews:
                                         Product Name  Number Of Reviews
23  Apple iPhone SE (White, 256 GB) (Includes EarP...               8161
55                      APPLE iPhone SE (Red, 128 GB)               8161
57                     APPLE iPhone SE (Black, 64 GB)               8161
53                    APPLE iPhone SE (Black, 128 GB)               8161
52                     APPLE iPhone SE (White, 64 GB)               8154
54                    APPLE iPhone SE (White, 128 GB)               8154
11  Apple iPhone XR (Coral, 128 GB) (Includes EarP...               6804
12  Apple iPhone XR (Black, 128 GB) (Includes EarP...               6796
13  Apple iPhone XR (White, 128 GB) (Includes EarP...               6796
10  Apple iPhone XR (Black, 64 GB) (Includes EarPo...               6796


## Visualize Star Rating Distribution

### Subtask:
Generate a histogram or bar chart to visualize the distribution of 'Star Rating' across all products. This will show the general sentiment of the ratings.


**Reasoning**:
To visualize the distribution of 'Star Rating', I will use `plotly.express.histogram` on the 'Star Rating' column of the `data` DataFrame and set the appropriate title.



In [40]:
figure = px.histogram(data, x='Star Rating', title='Distribution of Star Ratings')
figure.show()

## Visualize Sale Price Distribution

### Subtask:
Generate a histogram to visualize the distribution of 'Sale Price'.


**Reasoning**:
To visualize the distribution of 'Sale Price', I will use `plotly.express.histogram` on the 'Sale Price' column of the `data` DataFrame and add a descriptive title.



In [41]:
figure = px.histogram(data, x='Sale Price', title='Distribution of Sale Price')
figure.show()

## Final Task



## Summary:

### Data Analysis Key Findings

*   The average discount percentage across all analyzed Apple products is approximately 9.95%.
*   The top 10 iPhones by 'Number Of Ratings' are primarily dominated by various models of the Apple iPhone SE (e.g., Apple iPhone SE (White, 256 GB), APPLE iPhone SE (Red, 128 GB)) and Apple iPhone XR, with rating counts around 95,909 for SE models and 79,500 for XR models.
*   Similarly, the top 10 iPhones by 'Number Of Reviews' largely consist of Apple iPhone SE (256GB, 128GB, and 64GB variants) and Apple iPhone XR (128GB and 64GB) models, with review counts ranging from 6,796 to 8,161.
*   The distribution of 'Star Rating' was successfully visualized using a histogram, providing an overview of customer sentiment.
*   The distribution of 'Sale Price' was successfully visualized using a histogram, illustrating the price ranges of the products.